MovieTime/MovieTime/Core/MTCellProtocol.swift

```diff
@@ -0,0 +1,43 @@
+//
+//  MTCellProtocol.swift
+//  MovieTime
+//
+//  Created by Enrique Melgarejo on 09/01/18.
+//  Copyright © 2018 Choynowski. All rights reserved.
+//
+
+import UIKit
+
+protocol MTCellProtocol: class {
+    static var identifier: String { get }
+}
+
+extension MTCellProtocol where Self: UITableViewCell {
+    static var identifier: String {
+        return String(describing: self)
+    }
+}
+
+extension MTCellProtocol where Self: UITableViewHeaderFooterView {
+    static var identifier: String {
+        return String(describing: self)
+    }
+}
+
+extension MTCellProtocol where Self: UICollectionViewCell {
+    static var identifier: String {
+        return String(describing: self)
+    }
+}
+
+protocol MTSetupCellProtocol: MTCellProtocol {
+    associatedtype ModelType
+
+    func setup(with model: ModelType?)
+}
+
+protocol MTViewModelSetupCellProtocol: MTCellProtocol {
```

Comment:
how do you envision this protocol being reused in a future version of this app?

Reply:
Actually I intend not to use this protocol in the future. When I started the development, I thought about using protocols to connect the ViewModel in the cell, but protocols can only be used as a generic constraint and this one has a Self requirement.
In future versions, I will still use a protocol to load the ViewModel into cells (Views, etc), however I will use a class to implement the DataSource protocols (*UICollectionViewDataSource* and *UITableViewDataSource)* and make this connection.
Something like this:

```
let model = items[indexPath.row]
let cell = collectionView.dequeueReusableCell(withReuseIdentifier: model.assemble.identifier, for: indexPath)

model.assemble.update(view: cell) 

return cell

-------------------------------

func update(view: UIView) {
    (view as? View)?.load(viewModel: viewModel) // The cell receives a viewModel reference.
}
```